File: cloudflare-worker/src/push-provider-senders/ntfy.ts
import type { NotificationMessage, ProviderConfig, PushResult } from "../types";
import { fetchWithTimeout } from "../rocom-client";
import { providerErrorResult, resultFromParsedResponse } from "../push-http";
import { providerConfigText } from "./common";

export async function sendNtfy(
  provider: ProviderConfig,
  message: NotificationMessage,
  timeoutSec: number
): Promise<PushResult> {
  const baseUrl = providerConfigText(provider, "base_url").replace(/\/$/, "");
  const url = `${baseUrl}/${provider.config.topic}`;
  const headers: Record<string, string> = {
    Title: message.title,
    Markdown: "yes",
  };
  for (const [cfgKey, headerName] of [
    ["priority", "Priority"],
    ["tags", "Tags"],
  ] as const) {
    const v = providerConfigText(provider, cfgKey);
    if (v) headers[headerName] = v;
  }
  const token = providerConfigText(provider, "token");
  if (token) headers["Authorization"] = `Bearer ${token}`;

  try {
    const resp = await fetchWithTimeout(
      url,
      {
        method: "POST",
        headers,
        body: message.markdown,
      },
      timeoutSec
    );
    const text = await resp.text();
    return resultFromParsedResponse(
      provider,
      resp,
      {},
      text,
      new Set([null, undefined])
    );
  } catch (err) {
    return providerErrorResult(provider, err);
  }
}

File: src/roco_serverchan_notifier/push_provider_senders/ntfy.py
from __future__ import annotations

from ..push_http import HttpSession, result_from_status
from ..push_models import NotificationMessage, ProviderConfig, PushResult
from .common import provider_config_text


def send_ntfy(
    provider: ProviderConfig, message: NotificationMessage, session: HttpSession, timeout: int
) -> PushResult:
    base_url = provider_config_text(provider, "base_url").rstrip("/")
    url = f"{base_url}/{provider.config['topic']}"
    headers = {
        "Title": message.title,
        "Markdown": "yes",
    }
    for name, header in (("priority", "Priority"), ("tags", "Tags")):
        value = provider_config_text(provider, name)
        if value:
            headers[header] = value
    token = provider_config_text(provider, "token")
    if token:
        headers["Authorization"] = f"Bearer {token}"
    response = session.post(
        url,
        data=message.markdown.encode("utf-8"),
        headers=headers,
        timeout=timeout,
    )
    return result_from_status(provider, response)
